modules/consensus/synchronize_test.go

```diff
@@ -311,3 +315,197 @@ func TestSendBlocksBroadcastsOnce(t *testing.T) {
 		}
 	}
 }
+
+// TestRPCSendBlocks tests that the SendBlocks RPC correctly synchronizes with
+// peers.
+func TestRPCSendBlocks(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	// Create the "remote" peer.
+	cst, err := blankConsensusSetTester("TestRPCSendBlocks - remote")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cst.Close()
+	// Create the "local" peer.
+	//
+	// We create this peer manually (not using blankConsensusSetTester) so that we
+	// can connect it to the remote peer before calling consensus.New so as to
+	// prevent SendBlocks from triggering on Connect.
+	testdir := build.TempDir(modules.ConsensusDir, "TestRPCSendBlocks - local")
+	g, err := gateway.New(":0", filepath.Join(testdir, modules.GatewayDir))
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = g.Connect(cst.cs.gateway.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+	cs, err := New(g, filepath.Join(testdir, modules.ConsensusDir))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	blocksToMine := []types.BlockHeight{
+		0,                    // Try syncing when peers are at the same current block.
+		MaxCatchUpBlocks + 1, // There was a bug that caused SendBlocks to be one block behind when its peer was ahead by (k * MaxCatchUpBlocks) + 1 blocks.
+		0,                    // Sync again to let cs catch up in case it is behind.
+		MaxCatchUpBlocks + 2,
+		0,
+		MaxCatchUpBlocks,
+		0,
+		MaxCatchUpBlocks - 1,
+		0,
+		MaxCatchUpBlocks - 2,
+		0,
+		2*MaxCatchUpBlocks + 1,
+		0,
+		1,
+		2,
+		10,
+		11,
+		12,
+	}
+
+	for _, n := range blocksToMine {
+		for i := types.BlockHeight(0); i < n; i++ {
+			b, err := cst.miner.FindBlock()
+			if err != nil {
+				t.Fatal(err)
+			}
+			err = cst.cs.managedAcceptBlock(b)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		err = cs.gateway.RPC(cst.cs.gateway.Address(), "SendBlocks", cs.threadedReceiveBlocks)
```

Comment:
Would be greatly reassuring if you could drop a sanity check here to make sure that the consensus set hasn't actually caught up yet.

It's also worth me pointing out that when DEBUG mode is running, the cost of validating a block is linear in the total number of blocks (because the consensus set has some very aggressive sanity checks when DEBUG is enabled), so you might want to consider resetting the original consensus set every test as well. (otherwise you'll get to a height of 50+, which starts to slow testing down a lot).